packages/api/src/controllers/animalController.js

```diff
@@ -0,0 +1,151 @@
+        return res.status(500).json({
+          error,
+        });
+      }
+    };
+  },
+
+  addAnimals() {
+    return async (req, res) => {
+      const trx = await transaction.start(Model.knex());
+
+      try {
+        const { farm_id } = req.headers;
+        const result = [];
+
+        if (!Array.isArray(req.body)) {
+          await trx.rollback();
+          return res.status(400).send('Request body should be an array');
+        }
+
+        for (const animal of req.body) {
+          if (!animal.identifier && !animal.name) {
+            await trx.rollback();
+            return res.status(400).send('Should send either animal name or identifier');
+          }
+
+          if (!animal.default_type_id && !animal.custom_type_id) {
+            await trx.rollback();
+            return res.status(400).send('Should send either default_type_id or custom_type_id');
+          }
+
+          if (animal.default_type_id && animal.custom_type_id) {
+            await trx.rollback();
+            return res
+              .status(400)
+              .send('Should only send either default_type_id or custom_type_id');
+          }
+
+          if (animal.default_breed_id && animal.custom_breed_id) {
+            await trx.rollback();
+            return res
+              .status(400)
+              .send('Should only send either default_breed_id or custom_breed_id');
+          }
+
+          if (animal.default_type_id) {
+            const defaultType = await DefaultAnimalTypeModel.query().findById(
+              animal.default_type_id,
+            );
+
+            if (!defaultType) {
+              await trx.rollback();
+              return res.status(400).send('default_type_id has invalid value');
+            }
+          }
+
+          if (animal.custom_type_id) {
+            const customType = await CustomAnimalTypeModel.query().findById(animal.custom_type_id);
+
+            if (!customType || customType.farm_id !== farm_id) {
+              await trx.rollback();
+              return res.status(400).send('custom_type_id has invalid value');
+            }
+          }
+
+          if (animal.default_breed_id) {
+            const defaultBreed = await DefaultAnimalBreedModel.query().findById(
+              animal.default_breed_id,
+            );
+
+            if (!defaultBreed || defaultBreed.default_type_id !== animal.default_type_id) {
+              await trx.rollback();
+              return res.status(400).send('default_breed_id has invalid value');
+            }
+          }
+
+          if (animal.custom_breed_id) {
+            const customBreed = await CustomAnimalBreedModel.query()
+              .whereNotDeleted()
+              .findById(animal.custom_breed_id);
+
+            if (
+              !customBreed ||
+              customBreed.farm_id !== farm_id ||
+              customBreed.default_type_id !== animal.default_type_id ||
+              customBreed.custom_type_id !== animal.custom_type_id
```

Comment:
These could compare `null` and `undefined`, and an error is returned unexpectedly.
I was trying to create an animal with a custom type and a custom breed and didn't send `default_type_id`. `customBreed.default_type_id` was `null`, but `animal.default_type_id` was `undefined`, so I got a 404. `default_type_id` (or `custom_type_id`) null should be included in the request body?

Reply:
It shouldn't! Great catch, I've updated the logic and modified the test to assert it

Reply:
Ohhh this is interesting, so apparently this is caused by the animal having both a name and an identifier and that breaking the "oneOf" in the model. I changed it to "anyOf" now and it seems to be working fine, can you test again please?